backends/dpdk/dpdkArch.cpp

```diff
@@ -0,0 +1,920 @@
+    bin_expr = new IR::Shl(left_root, right_root);
+  } else if (bin->is<IR::Shr>()) {
+    bin_expr = new IR::Shr(left_root, right_root);
+  } else if (bin->is<IR::Equ>()) {
+    bin_expr = new IR::Equ(left_root, right_root);
+  } else if (bin->is<IR::LAnd>()) {
+    bin_expr = new IR::LAnd(left_root, right_root);
+  } else if (bin->is<IR::Leq>()) {
+    bin_expr = new IR::Leq(left_root, right_root);
+  } else if (bin->is<IR::Lss>()) {
+    bin_expr = new IR::Lss(left_root, right_root);
+  } else if (bin->is<IR::Geq>()) {
+    bin_expr = new IR::Geq(left_root, right_root);
+  } else if (bin->is<IR::Grt>()) {
+    bin_expr = new IR::Grt(left_root, right_root);
+  } else if (bin->is<IR::Neq>()) {
+    bin_expr = new IR::Neq(left_root, right_root);
+  } else {
+    std::cerr << bin->node_type_name() << std::endl;
+    BUG("not implemented");
+  }
+  stmt.push_back(new IR::AssignmentStatement(root, bin_expr));
+  return false;
+}
+
+bool ExpressionUnroll::preorder(const IR::MethodCallExpression *m) {
+  auto args = new IR::Vector<IR::Argument>;
+  for (auto arg : *m->arguments) {
+    sanity(arg->expression);
+    visit(arg->expression);
+    if (!root)
+      args->push_back(arg);
+    else
+      args->push_back(new IR::Argument(root));
+  }
+  root = new IR::PathExpression(IR::ID(collector->get_next_tmp()));
+  decl.push_back(new IR::Declaration_Variable(root->path->name, m->type));
+  auto new_m = new IR::MethodCallExpression(m->method, args);
+  stmt.push_back(new IR::AssignmentStatement(root, new_m));
+  return false;
+}
+
+bool ExpressionUnroll::preorder(const IR::Member *) {
+  root = nullptr;
+  return false;
+}
+
+bool ExpressionUnroll::preorder(const IR::PathExpression *) {
+  root = nullptr;
+  return false;
+}
+
+bool ExpressionUnroll::preorder(const IR::Constant *) {
+  root = nullptr;
+  return false;
+}
+bool ExpressionUnroll::preorder(const IR::BoolLiteral *) {
+  root = nullptr;
+  return false;
+}
+
+const IR::Node *IfStatementUnroll::postorder(IR::IfStatement *i) {
+  auto code_block = new IR::IndexedVector<IR::StatOrDecl>;
+  ExpressionUnroll::sanity(i->condition);
+  auto unroller = new LogicalExpressionUnroll(collector);
+  i->condition->apply(*unroller);
+  for (auto i : unroller->stmt)
+    code_block->push_back(i);
+
+  auto control = findOrigCtxt<IR::P4Control>();
+  auto parser = findOrigCtxt<IR::P4Parser>();
+
+  for (auto d : unroller->decl)
+    injector.collect(control, parser, d);
+  if (unroller->root) {
+    i->condition = unroller->root;
+  }
+  code_block->push_back(i);
+  return new IR::BlockStatement(*code_block);
+}
+
+const IR::Node *IfStatementUnroll::postorder(IR::P4Control *a) {
+  auto control = getOriginal();
+  return injector.inject_control(control, a);
+}
+const IR::Node *IfStatementUnroll::postorder(IR::P4Parser *a) {
+  auto parser = getOriginal();
+  return injector.inject_parser(parser, a);
+}
+
+bool LogicalExpressionUnroll::preorder(const IR::Operation_Unary *u) {
```

Comment:
@gordonwucn please take a look. If this requires more change, perhaps in another PR?

Reply:
This idea makes sense to me. And it can make the code much clearer. Here are some detailed explanations:

For example, say we have a = b + c + d. First, we will need to catch the statement, in this case, a = b + c + d. Then we check all possible locations that may have complex expressions. In this case, it is the right side of the AssignmentStatement. We visit the right expression, in the meantime, we need an IndexVector<> to hold the extra statement that this complex expression may produce. Since we visit Expression in postorder, first we visit c + d, its immediate parent is another expression, so we create an AssignmentStatement with a temporary variable `t1` to hold the value and push this statement(`t1 = c + d`) to the IndexVector<> we just created. Then return this temporary variable as the new node. Second, we will visit `a + t1`. Since its parent is a statement, return this expression immediately. Then we will go back to the statement preorder/postorder(does not matter). If that IndexVector<> has extra statements, we create a BlockStatement, push all the statements in this IndexVector<> and push this statement(in this case `a = b + t1`) lastly. Return this BlockStatement as the new node. Basically, by following this pattern, we can easily unroll a complex Expression.

This requires some major changes. @hanw Can you add this to TODO?

Reply:
added.

Reply:
You guys will maintain this code, so I hope you will eventually fix this.